torch/distributed/rpc/api.py

```diff
@@ -66,10 +90,46 @@ def _wait_all_workers():
     terminate the RPC framework, and there is no guarantee that the RPC
     framework will work after this method returns.
     """
-    global _agent
+    if _agent is None:
+        return
 
-    if _agent:
-        _agent.join()
+    assert _ALL_WORKER_NAMES is not None, (
+        "`_ALL_WORKER_NAMES` is not initialized for `def _wait_all_workers`."
+    )
+    leader_worker_name = sorted(_ALL_WORKER_NAMES)[0]
+
+    self_worker_name = _agent.get_worker_info().name
+    assert self_worker_name not in _INTENDED_WORKER_NAMES, (
+        "Can not call `_wait_all_workers()` twice."
+    )
+
+    is_leader_worker = leader_worker_name == self_worker_name
+
+    # Phase 1: Followers send intents.
+    # All followers report intents to the leader.
+    if is_leader_worker:
+        _on_leader_follower_report_intent(self_worker_name)
+    else:
+        rpc_sync(
+            leader_worker_name,
+            _on_leader_follower_report_intent,
+            args=(self_worker_name,),
+        )
+
+    _PROCEED_SIGNAL.wait()
+
+    # Phase 2: Leader asks followers to proceed.
+    # Leader's signal is the first to be unblocked,
+    # after receiving all followers' intents.
+    if is_leader_worker:
+        # The leader sends out proceeed signals to all followers.
+        for follower_worker_name in _ALL_WORKER_NAMES - {leader_worker_name}:
+            rpc_async(follower_worker_name, _set_proceed_signal, args=())
+
+        # This is a hack to make the leader linger for a while to finish
+        # sending out the last message.
+        import time
+        time.sleep(0.2)
```

Comment:
Can we use rpc_sync above when the leader sends out proceed signals and get rid of this sleep?

Reply:
That does not work. Since followers are calling `::shutdown()` while followers are also enqueueing response work. Leader will wait and times out.

Reply:
Can we set the RPC timeout to be small, use rpc_sync(), catch any timeout errors and just log them and continue? This would ensure each node has atleast received the shutdown signal before we exit this method and we can also get rid of the sleep.

Reply:
@pritamdamania 
We don't have per-RPC timeout yet. If we change the global timeout setting, we may affect other user RPCs and introduce subtle bugs. So I prefer not modify timeout at the moment.

Reply:
Updating to that.